### PR TITLE
ServerControl improvements

### PR DESCRIPTION
**Pull request by Darkness6030**
- Made certain important variables public (and documented them)
- Made certain important variables configurable
- Made a better implementation of getNextMap()
- Added a way to set gameOverListener

- [x] I have read the [contribution guidelines](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md).
- [x] I have ensured that my code compiles, if applicable.
- [x] I have ensured that any new features in this PR function correctly in-game, if applicable.

### CODE REVIEW

File: core/src/mindustry/maps/Maps.java [hunk 1]
```diff
@@ -61,8 +63,19 @@ public void setMapProvider(MapProvider provider){
...
     }
 
+    /** Set the map that will override the next selected map. */
+    public void setNextMapOverride(Map nextMapOverride) {
+        this.nextMapOverride = nextMapOverride;
+    }
+
     /** @return the next map to shuffle to. May be null, in which case the server should be stopped. */
     public @Nullable Map getNextMap(Gamemode mode, @Nullable Map previous){
+        if (nextMapOverride != null) {
```
Comment (Anuken) [inline core/src/mindustry/maps/Maps.java:73]: incorrect formatting throughout this PR, see CONTRIBUTING.md - extra spaces

File: server/src/mindustry/server/ServerControl.java [hunk 2]
```diff
@@ -56,19 +56,53 @@ public class ServerControl implements ApplicationListener{
...
-    private @Nullable Map nextMapOverride;
-    private Interval autosaveCount = new Interval();
+    /** The file to which the logs are currently being written. */
+    public Fi currentLogFile;
+
+    /** Whether the server is currently waiting for the next map to be loaded. */
+    public boolean inGameOverWait;
+
+    /** The current timer for loading the next map. */
+    public Task lastTask;
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:66]: This *still* should not be public, and this documentation is lacking. If you need a method to cancel it, add that, but keep it private.

File: server/src/mindustry/server/ServerControl.java [hunk 3]
```diff
@@ -56,19 +56,53 @@ public class ServerControl implements ApplicationListener{
...
+
+    /** The last gamemode loaded on this server. */
+    public Gamemode lastMode;
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:69]: `@Nullable`

File: server/src/mindustry/server/ServerControl.java [hunk 4]
```diff
@@ -173,34 +207,8 @@ protected void setup(String[] args){
...
-
-                info("Selected next map to be @.", map.plainName());
-
-                play(true, () -> world.loadMap(map, map.applyRules(lastMode)));
-            }else{
-                netServer.kickAll(KickReason.gameover);
-                state.set(State.menu);
-                net.closeServer();
-            }
+            if(!inGameOverWait && gameOverListener != null)
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:210]: do not use brace-less if-statements on multiple lines, see guidelines

File: server/src/mindustry/server/ServerControl.java [hunk 5]
```diff
@@ -280,9 +287,11 @@ protected void setup(String[] args){
         toggleSocket(Config.socketInput.bool());
 
         Events.on(ServerLoadEvent.class, e -> {
-            Thread thread = new Thread(serverInput, "Server Controls");
-            thread.setDaemon(true);
-            thread.start();
+            if (serverInput != null) {
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:290]: formatting

File: server/src/mindustry/server/ServerControl.java [hunk 6]
```diff
@@ -1081,14 +1094,14 @@ public void run(){
...
-            Timer.schedule(lastTask, roundExtraTime);
+            Timer.schedule(lastTask, Config.roundExtraTime.num());
         }else{
             r.run();
         }
     }
 
-    private void logToFile(String text){
-        if(currentLogFile != null && currentLogFile.length() > maxLogLength){
+    public void logToFile(String text){
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:1103]: why make this public?

File: server/src/mindustry/server/ServerControl.java [hunk 7]
```diff
@@ -1048,11 +1056,16 @@ public void handleCommandString(String line){
...
-        nextMapOverride = map;
+        maps.setNextMapOverride(map);
+    }
+
+    public void play(Runnable run){
+        play(true, run);
     }
 
-    private void play(boolean wait, Runnable run){
+    public void play(boolean wait, Runnable run){
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:1068]: undocumented parameters and method, explain what it does if you're going to make it public

File: server/src/mindustry/server/ServerControl.java [hunk 8]
```diff
@@ -1048,11 +1056,16 @@ public void handleCommandString(String line){
         }
     }
 
+    @Deprecated
```
Comment (Anuken) [inline server/src/mindustry/server/ServerControl.java:1059]: needs explanation on why it's deprecated and what should be used instead

File: core/src/mindustry/net/Administration.java [hunk 9]
```diff
@@ -513,7 +513,9 @@ public static class Config{
         autosaveSpacing = new Config("autosaveSpacing", "Spacing between autosaves in seconds.", 60 * 5),
         debug = new Config("debug", "Enable debug logging.", false, () -> Log.level = debug() ? LogLevel.debug : LogLevel.info),
         snapshotInterval = new Config("snapshotInterval", "Client entity snapshot interval in ms.", 200),
-        autoPause = new Config("autoPause", "Whether the game should pause when nobody is online.", false);
+        autoPause = new Config("autoPause", "Whether the game should pause when nobody is online.", false),
+        roundExtraTime = new Config("roundExtraTime", "Time before loading a new map after the gameover.", 12),
```
Comment (Anuken) [inline core/src/mindustry/net/Administration.java:517]: add "...in seconds"

**Comment by Darkness6030** [comment]
Ok, will do everything in ~9 hours, gotta go sleep XD

File: server/src/mindustry/server/ServerControl.java (hunk 3 above)
Comment (Darkness6030) [reply server/src/mindustry/server/ServerControl.java:69]: It can't be null

File: server/src/mindustry/server/ServerControl.java (hunk 6 above)
Comment (Darkness6030) [reply server/src/mindustry/server/ServerControl.java:1103]: For example, to make a custom logger

**Comment by Darkness6030** [comment]
@Anuken probably done

**Pull request merged**